### PR TITLE
User Authentication

### PR DESCRIPTION
**Pull request by neymarjimoh**
**What does this PR do?**
This PR authenticates the user. Tasks done include:
- User sign up
- User sign in
- User verification

**Description of Task to be completed?**
functionality if user forgets password and if logged in user wants to change password

**How should this be manually tested?**
- `git clone this repo`
- `cd product-investo-backend`
- `git checkout -b ft-crud-user`
- Install dependencies, `yarn`
- Run local server, `yarn dev`
- Test endpoints on `index.js` file on postman, cURL or postwoman.

**What are the relevant PT Story ID?**
[#173267277](https://www.pivotaltracker.com/story/show/#173267277)

### Screenshot Upon Verification
investveri

### Screenshot Upon Successful Login
investlog

### Screenshot Upon Successful Registration
investreg

### Screenshot- Mail notification for verification
invest

**Questions:**
- Do i write tests for each commit I make?

### CODE REVIEW

File: controllers/auth.controller.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        const generatedToken = cryptoRandomString({length: 15, type: 'url-safe'});
+        const user = new User({
+            email,
+            password: hashedPassword,
+            phoneNumber,
+            fullName,
+            verificationToken: generatedToken,
+        });
+        const savedUser = await user.save();
+        const verifyUrl = `http://localhost:${config.PORT}/api/v1/auth/verify-account/${savedUser.email}-${generatedToken}`;
```
Comment (Nicanor008) [inline controllers/auth.controller.js:28]: Get the URL host dynamically, don't type `localhost` manually.
Something like this:
```suggestion
        const verifyUrl = `http://ip/api/v1/auth/verify-account/${savedUser.email}-${generatedToken}`;
```

File: middlewares/auth.validation.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		body('password')
+			.not()
+			.isEmpty()
+			.isLength({ min: 8 })
+			.withMessage('Password must have at least 8 characters'),
+		body('phoneNumber')
+			.isLength({ min: 10, max: 15 })
+			.withMessage('Mobile Number must between 10 to 15 characters long')
+			.matches(/^[+-\d]+$/)
+			.withMessage('Mobile Number must be a valid Nigerian number'),
```
Comment (Nicanor008) [inline middlewares/auth.validation.js:22]: Make this global, not only based on your country

File: utils/sendMail.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      to,
+      subject,
+      text: msg,
+      html: msg, 
+    };
+
+    transporter.sendMail(mailOptions, (err, result) => {
+        if (err) {
+            console.log('error from sending mail >>>> ', err);
+            return res.status(500).json({
```
Comment (Nicanor008) [inline utils/sendMail.js:28]: This line will break the program and might stop(Fatal error) because `res` is not defined anywhere

**Comment by neymarjimoh** [comment]
Ok..I'll try and make the necessary changes.
I don't really understand the dns.lookup() method

File: controllers/auth.controller.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        const user = await User.findOne({ email, verificationToken: token });
+        if (!user) {
+			return res.status(statusCode.NOT_FOUND).json({
+				message: `Account ${email} doesn't exist. or ensure you enter the right url`,
+			});
+        }
+        // const updatedUser = await user.updateOne({ isVerified: true });
+        user.isVerified = true;
+        user.verificationToken = null;
+        const updatedUser = await user.save();
```
Comment (Nicanor008) [inline controllers/auth.controller.js:68]: I don't think it's a good idea to use `.save()` here, i.e. when updating a model. `.save()` will recreate a new function in the DB which is not good, use `.update()` instead.
Read more 
https://stackoverflow.com/questions/22278761/mongoose-difference-between-save-and-using-update

File: models/user.model.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
+const mongoose = require('mongoose');
+const Schema = mongoose.Schema;
+
+const userSchema = new Schema(
+    {
+        email: {
+            type: String,
```
Comment (Nicanor008) [inline models/user.model.js:7]: Try to register a user with this kind of email
`korir.nicanor@gmail.com`
Apparently it doesn't save the dot(.), kindly check on that. Most corporate emails do have dot

File: models/user.model.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            trim: true,
+            required: true,
+            unique: true
+        },
+        role: {
+            type: String,
+            default: 'user'
+        },
+        image: {
+            type: String
```
Comment (Nicanor008) [inline models/user.model.js:33]: Check on how to use multer to save images/files
https://www.npmjs.com/package/multer

**Comment by neymarjimoh** [comment]
Ok..I will try it and make the necessary changes

**Comment by neymarjimoh** [comment]
Should the user upload a picture on sign up??

File: routes/auth.route.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+const {
+    userSignUpValidationRules,
+    userSignInValidationRules,
+    validate,
+} = require('../middlewares/auth.validation');
+
+authRouter.post('/register', userSignUpValidationRules(), validate, register);
+authRouter.patch('/verify-account/:email-:token', activateAccount);
+authRouter.post('/login', userSignInValidationRules(), validate, login);
+authRouter.post('/forget-password', forgetPassword);
```
Comment (Nicanor008) [inline routes/auth.route.js:22]: how about 
```suggestion
authRouter.post('/forgot-password', forgotPassword);
```

**Comment by Nicanor008** [comment]
> Should the user upload a picture on sign up??

Yeah, on postman, you select type `form-data` and multer will do the rest. However if it willbe kind of complex, you can leave it for now, then work on it later

**Comment by neymarjimoh** [comment]
Ok.
I'll do so.. should I add uploading of image on sign up by the user??

**Comment by neymarjimoh** [comment]
I'll try it out..
If I encounter any difficulty I'd get back to you

File: models/user.model.js (hunk 6 above)
Comment (Nicanor008) [reply models/user.model.js:34]: Good job on this so far

**Pull request merged**